typing/parmatch.ml

```diff
@@ -2165,15 +2184,33 @@ let rec matrix_stable_vars rs = match rs with
     let stables_in_varsets = reduce (List.map2 IdSet.inter) rows_varsets in
 
     (* The stable variables are those stable at any position *)
-    List.fold_left IdSet.union IdSet.empty stables_in_varsets
-| rs ->
-   let submatrices = split_rows (simplify_first_col rs) in
-   let submat_stable = List.map matrix_stable_vars submatrices in
-   (* a stable variable must be stable in each submatrix;
-      if the matrix has at least one row, there is at least one submatrix *)
-   reduce IdSet.inter submat_stable
-
-let pattern_stable_vars p = matrix_stable_vars [{varsets = []; row = [p]}]
+    Vars (List.fold_left IdSet.union IdSet.empty stables_in_varsets)
+| ns, rows ->
+    let ns = simplify_first_col ns in
+    let rows = simplify_first_amb_col rows in
+    let extend_row columns r =
+      { r with row = columns @ r.row } in
+    let q0 = discr_pat omega rows in
+    match build_specialized_submatrices ~extend_row q0 rows with
+    | { default; constrs = [] } ->
+        (* the first column contains no head constructor;
+           they are all _ after simplification, so it can be dropped *)
+        let default_ns = List.map snd ns in
```

Comment:
That line looks fishy.

I spent a long time thinking about the coherence implications of that line, that is: clearly the `default_ns` (which seems like a bad name to me btw, `default_ns` is in no way the default matrix of `ns`) you produce can contain incoherences, but does it matter?
I'm still not quite sure but I believe it does and that we should care about the coherence of `ns`.
Here we probably want to build the specialized submatrices of ns and recurse on each of them (keep default for the other argument on each of the recursive call). This is just an intuition however and it will be easier to test out if this PR is rebased on top of mine.

In any case, I took some time to produce a convoluted example introducing incoherence in specific places and trying to see how the new implementation would behave on it. At which point I realized I could greatly simplify my test case to the following code (where all the columns are coherent at every step):
```ocaml
type a = A1 | A2

type 'a alg =
  | Val of 'a
  | Binop of 'a alg * 'a alg

let cmp (pred : a -> bool) (x : a alg) (y : a alg) =
  match x, y with
  | Val A1, Val A1 -> ()
  | Val x, _
  | _, Val x when pred x -> ()
  | _, _ -> ()
```
With your implementation the warning doesn't trigger on this example. I think it should.

I tried the idea I mentioned above (recursing on each of the specialized submatrices of `ns`) and it does warn on that specific example (`make core` then  `./boot/ocamlrun ./ocamlc -nostdlib -nopervasives -w +57 simple_amb.ml`) but `make world.opt` fails because of the `assert (ns = [])` a few lines above.
So my implementation is not quite right either, but you can probably get something along this line working.

Reply:
The diff of my attempt is the following btw:
```diff
diff --git a/typing/parmatch.ml b/typing/parmatch.ml
index 16949df57..3fbac8c43 100644
--- a/typing/parmatch.ml
+++ b/typing/parmatch.ml
@@ -2195,8 +2195,17 @@ let rec matrix_stable_vars ns rs = match ns, rs with
     | { default; constrs = [] } ->
         (* the first column contains no head constructor;
            they are all _ after simplification, so it can be dropped *)
-        let default_ns = List.map snd ns in
-        matrix_stable_vars default_ns default
+        let q0_ns = discr_pat omega ns in
+        begin match build_specialized_submatrices ~extend_row:(@) q0_ns ns with
+        | { default = default_ns ; constrs = [] } ->
+          matrix_stable_vars default_ns default
+        | { default = _; constrs = constrs_ns } ->
+          let submat_stable =
+            List.map (fun (_, sub_ns) -> matrix_stable_vars sub_ns default)
+              constrs_ns
+          in
+          List.fold_left stable_inter All submat_stable
+        end
     | { default = _; constrs } ->
         (* A stable variable must be stable in each submatrix.
 
```

Reply:
`(fun (_, sub_ns) -> matrix_stable_vars sub_ns default)` does not respect arity, as `sub_ns` may in general have more columns than `ns` (while `default` always has one less). In the non-default case here it is safer build the specialized matrix of `rs` against the head of each `constr_ns` group (so to drop the decomposition of `rs`), which suggests that we are really trying to split on both matrices at once instead of decomposing them independently. Also it is not clear whether it is right to forget about `default_ns` in that case (when `constrs_ns` is non-empty): default_ns has less content than any constrs_ns, but precisely having less negative information can lead to accepting more positive rows, and thus finding unstable variables.